services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_states.py

```diff
@@ -0,0 +1,107 @@
+"""
+States from Docker Tasks and docker Containers are mapped to ServiceState.
+"""
+import logging
+from typing import Dict, Set, Tuple
+
+from ...models.schemas.dynamic_services import ServiceState
+
+logger = logging.getLogger(__name__)
+
+TASK_STATES_FAILED: Set[str] = {"failed", "rejected"}
+TASK_STATES_PENDING: Set[str] = {"pending"}
+TASK_STATES_PULLING: Set[str] = {"assigned", "accepted", "preparing"}
+TASK_STATES_STARTING: Set[str] = {"ready", "starting"}
+TASK_STATES_RUNNING: Set[str] = {"running"}
+TASK_STATES_COMPLETE: Set[str] = {"complete", "shutdown"}
+
+
+TASK_STATES_ALL: Set[str] = (
+    TASK_STATES_FAILED
+    | TASK_STATES_PENDING
+    | TASK_STATES_PULLING
+    | TASK_STATES_STARTING
+    | TASK_STATES_RUNNING
+    | TASK_STATES_COMPLETE
+)
+
+
+# mapping container states into 4 categories
+CONTAINER_STATES_FAILED: Set[str] = {"restarting", "dead", "paused"}
+CONTAINER_STATES_PENDING: Set[str] = {"pending"}  # fake state
+CONTAINER_STATES_PULLING: Set[str] = {"pulling"}  # fake state
+CONTAINER_STATES_STARTING: Set[str] = {"created"}
+CONTAINER_STATES_RUNNING: Set[str] = {"running"}
+CONTAINER_STATES_COMPLETE: Set[str] = {"removing", "exited"}
+
+
+def _docker_task_state_to_service_state(state: str) -> ServiceState:
+    last_state = ServiceState.STARTING  # default
+
+    if state in TASK_STATES_FAILED:
+        last_state = ServiceState.FAILED
+    elif state in TASK_STATES_PENDING:
+        last_state = ServiceState.PENDING
+    elif state in TASK_STATES_PULLING:
+        last_state = ServiceState.PULLING
+    elif state in TASK_STATES_STARTING:
+        last_state = ServiceState.STARTING
+    elif state in TASK_STATES_RUNNING:
+        last_state = ServiceState.RUNNING
+    elif state in TASK_STATES_COMPLETE:
+        last_state = ServiceState.COMPLETE
+
+    return last_state
+
+
+def _docker_container_state_to_service_state(state: str) -> ServiceState:
+    last_state = ServiceState.STARTING  # default
```

Comment:
why not creating a static ``dict`` that maps input/output? probably faster and built once

Reply:
Finally figured out a clean way to write this.